### ISSUE TITLE
Support for .tt templates in .kproj files?

### ISSUE THREAD

**Issue by kevinkuszyk** [aspnet/Home#272]
Is there a plan to include support for `.tt` files in `.kproj` files in vnext?

I just tried adding T4MVC to an MVC6  project but the `.tt` files didn't get copied into the project.  

When I manually copied the `.tt` into the project folder, it showed in in the solution explorer, but the option to run it was missing.

**Comment by Eilon** [comment]
No current plans to support it.

**Comment by kevinkuszyk** [comment]
Thanks for the quick reply.

Fyi. - I just posted this on the user voice page too:
https://aspnet.uservoice.com/forums/252111-asp-net-vnext/suggestions/6970908-support-t4-templates-in-kproj-in-visual-studio

**Comment by wwwlicious** [comment]
+1 

Until MVC changes from using awful magic strings for routes, partials etc T4MVC has been invaluable for refactor friendly, view compile-time checking.

@Html.ActionLink("Home", "Index", "Home") == "urgh"
@Html.ActionLink("Home", MVC.Home.Index()) == "win"

**Comment by Eilon** [comment]
Indeed, you'll get no argument from anyone on the team that hard-coding repeated strings all over the place is rather... lame. OK actually it's *super* lame.

As part of this discussion I think it's important to share my thoughts as to why TT probably isn't going to be the long-term solution. First, it's pretty much a VS-only, design-time-only solution. People have done hacky workarounds to run TT files from the command-line but it's not a supported pattern. And in the world of ASP.NET 5, being a Windows-only VS-only solution just isn't going to cut it. I have to be able to use the feature when part of my team might be running Ubuntu and using vim to work on their part of the project. Secondly, editing support for TT files in VS is rather weak. There have been some reasonable 3rd party solutions, but nothing in-box.

So then what do we choose instead? I call my next witness, the Roslyn C# compiler. Roslyn lets you do neat things. Roslyn lets you do some pretty *amazing* things. One of those amazing things with the way ASP.NET 5 uses Roslyn is that it lets you at *compile time* (which is also potentially at *run time* (well, *load time*, really)) process the syntax trees, generate some code of your own, and then let the world keep turning.

This is how ASP.NET 5 does [assembly neutral interfaces](http://davidfowl.com/assembly-neutral-interfaces/) (ANIs). ANIs are implemented in ASP.NET 5 by having a scanner that looks for certain attributed types and *deletes* their syntax trees, and, instead, creates new tiny assemblies with just those types.

Now imagine a world where you could do the same functionality as you had with T4MVC, but potentially far more robust (you have Roslyn, after all), far faster (you already have all the syntax trees parsed), cross-platform (Windows, *nix, OS X, etc.), and can generate code hopefully more easily (well, with Roslyn that isn't always the case, but still).

So, I'd love to hear thoughts on a solution that heads in that direction. Anyone?

**Comment by kevinkuszyk** [comment]
I was thinking about this some more last week and came to the conclusion cross platform support might be the blocker and Roslyn the solution.

I opened this issue because I don't want to loose access to the strongly typed action name, view names etc. we get from T4MVC.

If there is better cross platform solution then I'd go for that.

Do you see a Roslyn based T4MVC replacement shipping as part of MVC 6 or in a future version, or is there still space for a community lead extension?

**Comment by Eilon** [comment]
The team doesn't have any immediate plans to build this out, so I think a community-organized solution is the best place to start. But, of course, the team would love to hear about any roadblocks or issues found while building such a solution because we would very much like to unblock any such issues.

**Comment by kevinkuszyk** [comment]
Thanks @Eilon.  I'll try to build something, but I'm not sure where to start.  I have two big questions to begin with, but I'm not sure where to post them:

1. Is there a `project.json` hook we can use to invoke our code generation?  Or should we use the `BeforeBuild` target in the `.krpoj` file.
2. How to use Roslyn "reflection" to find all the controller classes in a project?  I've looked through the samples in the [Roslyn](https://github.com/dotnet/roslyn) repro, but I can't see an example of loading a project and finding types in it.

**Comment by davidfowl** [comment]
> 

You can follow the same pattern as the RazorView preprocessor https://github.com/aspnet/Mvc/blob/dev/samples/MvcSample.Web/Compiler/Preprocess/RazorPreCompilation.cs. In the target project, the file needs to go into compiler/preprocess/*.cs.

> How to use Roslyn "reflection" to find all the controller classes in a project? I've looked through the samples in the Roslyn repro, but I can't see an example of loading a project and finding types in it.

You need to play with the API and learn how to use it. Not sure where you can find better examples but they do exist somewhere

**Comment by mcquiggd** [comment]
This really is not good enough.

There is huge investment in T4 throughout the .Net community, focusing directly on MVC applications from T4 scaffolding of custom project items, through utilities, to entire frameworks. There is an implementation of T4 for Mono; T4 is open source.

We have T4 templates for generating TypeScript, KnockOut Models, just about everything you could imagine that is related to web development.

While Visual Studio 2015 increases support for T4, the ASP.Net vNext team decided to quietly drop it completely. And to add insult to injury, the reasoning mentioned is because 'T4 would not work for my team members that use Ubuntu'. Comments such as 'Just rewrite it in Roslyn', 'maybe the community will come up with something' are far from helpful.

I am seriously concerned at the direction the ASP.Net team are taking. I could just about suffer the pain of being forced to write tortuous Gulp scripts to compile SASS and put it in a location, bundling and minification, that *used* to be simple and straightforward, but to find out that the ASP.Net team prefer to focus on the needs of the tiny Ubuntu / Sublime text on a Mac user base, rather than their core audience of Visual Studio users, is maybe the last straw. 

Already .Net developers are looking at all these changes, and thinking 'why should I bother - if Microsoft doesn't even like its own ecosystem, I might as well become a Node developer'

**Comment by mcquiggd** [comment]
As a follow up, where is the best place to register opposition to this design decision? Can you recommend the relevant User Voice / blog / issue tracker for this, if you are not interested in listing to people?

**Comment by davidfowl** [comment]
We've been developing ASP.NET 5 so far by making so any solution we come up with is viable cross platform. It's a good principle and it helps us design things that can work well from the command line but that will also light up in visual studio.

We're looking at bringing back single file generators so things like T4 and resx will just work (within visual studio only of course). We'll likely have to find other solutions for problems like these when developing cross platform.

One of our devs even ported T4 to CoreCLR https://github.com/bricelam/TextTemplating.

Stay tuned.

**Comment by mcquiggd** [comment]
@davidfowl - thanks for the quick response. 

So, can you please clarify whether or not T4 will in fact be *officially supported* in ASP.Net MVC 5?

Seriously, I do get the cross-platform concept, and I see its importance for Microsoft. But you are losing existing developers due to these kind of decisions. Check out the threads in even the .Net groups on LinkedIN, for evidence that your core audience is having doubts and looking around.

a) Its a good principal - for runtime if people want to host on Linux. But - wont someone 'think of the developers..?' ;-) Don't make the design decisions for the lowest common denominator (a text editor). Your ace card is precisely *Visual Studio*. If T4 is not available for sublime text etc, then let their 'community' add it if they want. Don't just remove it for the majority of us, who need Visual Studio for the other one thousand and one things it does better in the Microsoft ecosystem than a text editor.

b) At least provide a transition period - and pre-warn people that if they expect to be able to move to MVC vNext, they will be losing T4 support. This has not been made public, amid all the celebrations of adding bower, NPM, Grunt, Gulp, etc etc.

c) I would seriously think it would be easier and simpler to build a T4 host for the tiny proportion of Linux and Mac users who would need it.

**Comment by davidfowl** [comment]
> So, can you please clarify whether or not T4 will in fact be officially supported in ASP.Net MVC 5?
I think you mean MVC 6 not 5.

T4 hasn't been removed per se, our project system just doesn't support single file generators at the moment (the little Custom tool icon on file properties). We're actively investigating the work that it will take to bring this back.

As for T4 on .NET Core? That's not a thing but it won't matter unless you're trying to port T4 itself to run on the new CLR. If you're only concerned with your existing Visual Studio workflow then it won't make a difference.

Things are generally removed by default then brought back when we figure out how to make things work well everywhere (more people use sublime than you think :smile:). That's not a hard and fast rule that can't be broken though. 

I'll update this thread when we have more information about our single file generator story.

/cc @sayedihashimi @BillHiebert

**Comment by mcquiggd** [comment]
@davidfowl 

Thanks again...

To reiterate, my personal interest (and probably also for the hundreds of thousands who don't know about the lack of T4 support as it stands today), is simply that we need to keep T4 available and functional within ASP.Net MVC5, *as it is* within other Visual Studio 2015 project types, for existing users so they can continue to leverage the time and investment they have made, in a technology that Microsoft is still promoting outside the ASP.Net vNext team.

I know a lot of people use Sublime - I do too - but just like 99% of the others, I do not use it for .Net ;-)

Off topic - seriously, if Microsoft want to be relevant client side, invest some love in http://Aurelia.io. 

It is an opportunity to leapfrog Google (Angular), Facebook (React), and Twitter (Bootstrap) and become a name in web dev rather than be viewed as legacy. If you do not have a presence in the client side landscape, people will continue to go straight from Angular etc to Node (ironically using Sublime or WebStorm which is becoming the alternative to Visual Studio) and bypass Microsoft completely. It is this huge gap in Microsoft's toolset that is leading you to have to play catch up rather than leading. 

Don´t forget your existing developers chasing open source users who already view Microsoft as an enemy.

**Comment by davidfowl** [comment]
> Well, T4 does not work in ASP.Net MVC 5 preview projects ;)

ASP.NET 5 -> MVC 6

> To reiterate, my personal interest (and probably also for the hundreds of thousands who don't know about the lack of T4 support as it stands today), is simply that we need to keep T4 available and functional within ASP.Net MVCx Preview, as it is within other Visual Studio 2015 project types, for existing users so they can continue to leverage the time and investment they have made, in a technology that Microsoft is still promoting outside the ASP.Net vNext team.

That's well understand. Although you'll still have to tweak the generated content as the frameworks have changed.

**Comment by mcquiggd** [comment]
@davidfowl  Tweaking is not a problem, compared to not having the option at all...

Thanks for your feedback - I will shut up, for now :-)

**Comment by mcquiggd** [comment]
P.S can we please re-open or create a new issue for this, as it is obviously not closed.

**Comment by Perambulator** [comment]
+1 for full T4 support or at least design time

This came as a surprise to me because I assumed that T4 would be added and this was just in beta. I cant imagine why T4 would be seen as being redundant as it is so heavily used by existing tools and commercial products. Say it aint so!

**Comment by rogerhendriks** [comment]
+1

**Comment by aukgit** [comment]
Well in MVC 2 demos Microsoft repeatedly mention to use T4 and now they putting it in hold. For few months I have seen that every new thing that Microsoft created , destroyed the previous one. When EntityFramework 6 released, it was a whole new thing , they move namespaces from .NET to EntityFramework 6 which cause lot of problems with existing projects. Idea was to have a standalone product. Okay good, now in EF7 they are going to break everything and it almost a new thing and many support(edmx , database first approach[not sure]) will not be there too. Which begs a serious question "does Microsoft build anything with their technology?" .

I like T4 and I seriously don't like Microsoft.

**Comment by PericlesSouza** [comment]
+1 for T4 please

**Comment by guardinfo-spam** [comment]
+1 for T4 please. there's a lot of time invested in T4 and frankly if Ubuntu users don't like it then they don't have to use it. Those of us who made extensive use of T4 templates need them to work as they always did. Can the team please consider this seriously. If they only work on windows then so be it. Your vast majority of users is on windows and we're simply offering this world to other OSes. 

Dropping important parts of the system to please a specific platform is not a good idea and it will end up hurting you in the long term. You can't drop support just like that and you cannot seriously expect the community to do your job for you. 

So please stop this thing with Ubuntu users or whatever other platform you want and think about your current users. Don't kick us all out in your quest to gain new users. You need to keep your existing developer base happy. If you do drop it, you will basically make us all think that you can stop support for anything very quickly which will more than likely alienate a very large base of your current users.

Do you really want this to happen ?

**Comment by josher565** [comment]
+1 for T4
T4 has been flying under the radar for a long time, it typically gets the last consideration in the past and it probably is here too. That's ok, but having a generation system is really handy for a lot of uses. I suspect T4 competes with other directions in the current project (yeoman support, roslyn). That's ok too, just so long as we know how/when to use competing solutions and how it supports the old world of T4. I suspect supporting T4 will be simpler. Its probably a good candidate for OSS to let the voices here and others shape and see if it's as important as it's suspected. It probably hasn't found it's ultimate form quite yet and getting Linux feedback (in code or otherwise) would be interesting. Is there such a facility in the LAMP/POSIX stack other than npm tools?

**Comment by mcquiggd** [comment]
T4 is already open source, and is supported in IDEs such as SharpDevelop, MonoDevelop.

It's been available on Mono for about 5 years.

Yeoman doesn't cut it - try generating TypeScript definitions of your .Net View Models with it. Easy in T4. 

T4 is used throughout the .Net stack, but we will end up with MVC being the odd man out.

**Comment by mcquiggd** [comment]
+1 for T4 - in case the team had not guessed by now ;-)

One other advantage of T4 is that you can output any file type you want - JavaScript, TypeScript, documentation, you have access to ENVDte for solution and project manipulation in Visual Studio, you can use CodeDom, FileCodeModel, or Rosyln inside it, and output whatever you want. It is already there and does not need to be recreated. Its a very flexible container for performing code generation within every aspect of .Net, from creating DbContexts, to DTOs with data annotations, to WCF service interfaces, to custom Controllers, MVC Views, of WPF.

It can act as a bridge between server side MVC view models and client side JavaScript frameworks (as I highlighted in my link from the Aurelia project https://github.com/aurelia/framework/issues/72). This is how I build Knockout models, Angular models, Aurelia models - in TypeScript via T4 based on meta data. 

And it can do this with strong typing and intellisense - right now.

Given time I could migrate the code in my T4 templates to use Rosyln rather than ENVDte for obtaining meta data, but there is no coherent story yet on this subject, and I am wary of investing time to do this, if it gets dumped without any warning.

**Comment by PericlesSouza** [comment]
A question about the comment that resx files are not currently supported "We're looking at bringing back single file generators so things like T4 and resx will just work (within visual studio only of course)"

Does this mean that third party libraries like Humanizer, NodaTime and others will not work in MVC6? They use resx files to store their data.

**Comment by TareqAtiek** [comment]
+ 1

Seems to be ivory tower approach, not real world usage. Too much focus by the team on projects like OmniSharp and only supporting what can be implemented in text editors used by a few people with loud voices. Very conflicting messages from Microsoft when releasing Community edition of Visual Studio but dropping features from their code frameworks to support everything that is not Visual Studio. There seems to have been no planning for this.

**Comment by mcquiggd** [comment]
Here is one suggested course of action, based on a few assumptions:

**Assumptions**

1. Roslyn is the future of .Net *code* generation. It will act as the cross platform way to dynamically create *code*, inspect *code*, etc.

2. Rosyln currently lacks the capabilities of T4 for *file* generation (files can consist of *any content type* and are not limited to *code*, which presumably is why T4 stands for Text Template Transformation Toolkit).

3. T4 can use Roslyn for extracting meta data from code, and offer the ability to output that in any content type.

**Suggestion**

a) T4 is retained as is for MVC6 (and resx support) to allow a managed upgrade path for existing .Net developers.

b) A discussion about the requirements of code generation is held, including input from the developer community and the other Microsoft teams that use it (e.g. Entity Framework, Visual Studio), which lead to a specification for a best-practice solution. No more hacking it together in isolation.

c) Whatever is the agreed solution, should be used throughout the .Net framework. Not bits of T4 here and then missing in other projects types that use something else.

d) Clearly notify people of the outcome - if T4 is to be end of life, and Roslyn extended to take its place, then at least warn us and give us time to prepare. Give guidance on how to use the technology.

e) The relevant Microsoft teams plan and implement the agreed solution, and synchronise a release schedule.

f) In the meantime, a list of features that will not be implemented in MVC6 needs to be created; hacking it together and letting .Net developers find out by accident is not good enough for a company of Microsoft's ability and resources. 
 

Give it a catchy name - Code and Content Creation Components -  'C4', powered by Roslyn, whatever. 

But please, I beg you, avoid the temptation to throw in open source components that have a half-life of 3 months at the expense of technologies that have been around for 8 years. Don't be scared to use Microsoft technologies, improve them, and make them best of breed.

**Comment by davidfowl** [comment]
I'm not sure what else to say, we're bringing back single file generators at design time.

**Comment by mcquiggd** [comment]
@davidfowl  

Well, thanks for finally confirming it; that is good news indeed.

Enjoy //build ;)

**Comment by TareqAtiek** [comment]
@davidfowl thank you for the reassurance that you listen to us.

**Comment by sayedihashimi** [comment]
Thanks for all the feedback here. We really appreciate it. I promise you that we are listening to your feedback. One of the reasons that we are more open is so that we can receive early feedback from the community. This is a great example of that.

Supporting T4 and Resource Files (.resx) require the same infrastructure, it's called Single File Generators. This is also used in other scenarios as well, so there is no doubt that it's an important feature to enable.

I wanted to let you know that it is on the backlog to be implemented, and comments here have moved it up in priority. I'm ***hoping*** that we can get this done in time for the next release of Visual Studio. With that said we still have a lot of work that we need to focus on before we are able to get to that. For example, some of the things we (_ASP.NET 5 Visual Studio project system team_) are/will be working on include.

 - Improving the quality and reliability of existing features
 - Improving performance for ASP.NET 5 projects
 - Ensuring that _Add Missing NuGet package_ works across all scenarios for ASP.NET 5
 - Improving the experience with project/solution errors
 - Improving the package restore experience

Honestly we will be spending a lot of time improving quality and reliability.

We want to get support for Single File Generators in the next release, but we cannot commit to it at this time. If we are not able to do so I assure you that you will see it in a future release after that.

Regarding the comments about not forgetting our core customers, we have not. We know that the vast majority of our users are in Visual Studio and on Windows. There are certain areas where we have purposefully implemented features that are specific to Visual Studio and Windows. One example of this is the new Web Publish process for ASP.NET 5. It's implemented in PowerShell, which is Windows specific. We could have chosen to implement this in other ways, but we knew that most of you wouldn't have liked those other options.

Once again thanks for all the feedback, please do keep it coming.

**Comment by mcquiggd** [comment]
In return, although obviously I am speaking for myself, I am sure that others share the same sentiment when I say thanks to the team, for both your efforts, and your openness. Hope you guys had time to enjoy the response from the developer community today, following the //build presentations... ;-)

**Comment by Perambulator** [comment]
Microsoft need to be careful. The reaction from .Net developers has been "hey this is great". The reaction from Linux fans has been "It is Microsofts suicide note. Why would we bother?"

**Comment by mcquiggd** [comment]
To get back on topic...

@sayedihashimi 

Previously, in VS2013, we had the ability to customise templates used for Scaffolding Items, as described in your article here: 

http://blogs.msdn.com/b/webdev/archive/2013/12/26/how-to-customize-the-generated-files-from-the-new-scaffolded-item-dialog.aspx

Is there a 'story' or recommended practice for achieving similar functionality in VS 2015 and MVC6..?

**Comment by sayedihashimi** [comment]
@mcquiggd we will have a scaffolding story which is on par with what we had in VS2013 including the ability to override the templates in the local project.
We should discuss this in a separate issue, but we have two options regarding what tech we use to author the templates;
 - Use T4 like we did in previous versions
 - Use Razor

I am personally leaning towards using Razor because we have better editor support and when you compare the same template in Razor/T4 the Razor template is typically much smaller and easier to understand.

We actually already worked on this a bit, which you can read up on at http://blogs.msdn.com/b/webdev/archive/2014/08/23/how-to-customize-scaffolding-templates-for-asp-net-vnext.aspx. Fair warning: that is likely to change before ASP.NET 5 is available in final release so please do not make a big investment in it.

**Comment by mcquiggd** [comment]
@sayedihashimi 

Thanks - I was considering starting a new discussion, but did not want to pollute the issue tracker + the previous solution was T4 so it seemed related.. perhaps its best, if you are looking for feedback, that at a suitable time you could start the discussion topic with some scope laid out to keep the discussion on track?

I'll check the link in the meantime.

**Comment by weitzhandler** [comment]
Ditto!
Very essential.
I want to see `config.json`'s generated on design-time, like we were used so far. Highly demanding feature!

**Comment by bchavez** [comment]
I use [RazorGenerator](https://github.com/RazorGenerator/RazorGenerator) in `Template` mode that works really well as a substitute for T4. [VS Extension Here](https://visualstudiogallery.msdn.microsoft.com/1f6ec6ff-e89b-4c47-8e79-d2d68df894ec).

Highly recommend checking it out.

**Comment by mcquiggd** [comment]
This reminds me ...

@sayedihashimi - I have not seen a new dicussion about the future of T4 / its replacement in a  future release of MVC; I hope this is not forgotten, but I understand that the team have priorities.

@bchavez  - Thanks for bringing this to our attention; I will look into this in more detail.

**Comment by sayedihashimi** [comment]
@mcquiggd no, we have not forgotten about it. We are actually working on the underlying support now. FYI we are going to be posting a roadmap soon in https://github.com/aspnet/tooling.

**Comment by claysmith** [comment]
For now I can just add a console app to my vnext project and set post-build to xcopy my generated t4 file. A bit of a hack short term solution.

**Comment by mcquiggd** [comment]
@claysmith 

Yes I do a similar thing, in terms of using a separate non ASP.Net 5 project, and placing the outputs of the T4 into the ASP.Net 5 project (not using a build step). Very hacky, but it has allowed me to at least continue. 

A number of useful MVC5 compatible Nuget packages exist which I cannot use with MVC6 at the moment - I am hoping we have an answer before the authors give up and move elsewhere.

**Comment by davidfowl** [comment]
Those packages will still have to be updated.

**Comment by mcquiggd** [comment]
That's my point - I can make my own T4 stuff work at the moment, but not other people published packages. Hopefully they get a roadmap for creating MVC6 compatible versions, where applicable.

**Comment by rpokrovskij** [comment]
I agree to remove Razor  because we will get better instruments instead (we will work in another paradigm where there are no "server-side only"  templates - and I appreciate it). But removing T4 what you are giving to us? We use T4 to generate code (typescript, html, various templates) from EF6 model? How we could use Roslyn for it (to generate ts,html)? How we could use yeoman to analize EF6 model? And again we can't migrate to yeoman today exactly because we will need tomorrow migrate to Roslyn. And we have a lot of code. So please keep T4 and T4 scaffolding  working (at least from command line). It is not about the web client-server architecture it is just a quick web generation from EF6. Or propose anything that can replace current T4 scaffolding for us.

**Comment by BillHiebert** [comment]
Note that T4 is supported in the RC release of ASP.NET 5 if you have Visual Studio 2015 Update1 installed. It uses the same custom tool mechanism that csproj projects use.

**Comment by dls314159** [comment]
@BillHiebert I'm not sure if you mean that you can import and use the Microsoft.TextTemplating.targets from the Modeling SDK for Microsoft Visual Studio 2015, but I found out that you can do that in RC 1.

**Comment by totht91** [comment]
I created a tool for asp.net 5 to generate t4 templates in design time (in linux or mac too). https://github.com/totht91/TextTemplating

**Comment by MaxHorstmann** [comment]
@BillHiebert great news, can't get T4MVC to work with ASP.NET 5 yet though, see http://stackoverflow.com/questions/34399939/make-t4mvc-work-with-asp-net-5

**Comment by fenixbrood** [comment]
I have been thinking about using Roslyn analysers to get the same functonality.

**Comment by fenixbrood** [comment]
It may not have the exact functionality. But it helps with the code generation for static strings in MVC.

**Comment by mredman** [comment]
+1 for T4 support in MVC6 and Class Library Packages.
T4's do seem to work in Class Library Packages (where we generate CRUD/boilerplate code from Entity Classes) but this only works when clicking "Run Custom Tool" on the .tt file, "Transform All T4 Templates" does not work.

Replacing this code generation is a lot of work and requires access to the class structure from the Entity classes, Is anything else that can do this at design time?

The shift to asp.net 5 MVC6 seems like a large jump, I am busy re-working a lot of code and apply a lot of new techniques to convert things, and having useful and dependent things discontinued and/or missing without alternatives just slows everything down.

**Comment by mcquiggd** [comment]
Hi all,

Personally I dislike Magic Strings intensely, and MVC is way too overloaded with them. I don't use T4MVC as I personally find that far too heavy for my needs.

So, I have uploaded an initial sample of automatically generating Controller Names, Action Names and View Names with both the **MVC5** and **MVC6** versions of MVC Boiler Plate.

https://github.com/Pro-Coded/MVCMagic

I will be adding more features over the next couple of days; let me know what you think... I want some feedback to see if there is an audience for this before releasing a NuGet probably next week.

David

**Comment by sayedihashimi** [comment]
FYI in RC1 we have added support for .tt files so I'll go ahead and close this. If there is more info here that I missed please let me know.

@mcquiggd thanks for the link, I'll check it out.

**Comment by mredman** [comment]
I don't believe its fully supported added given my experience of the last 2 days?

1) "Transform All T4 Template" in the Build menu doesn't work
2) Generating multiple files hangs VS when files are added to a class library package project.

**Comment by kevinkuszyk** [comment]
@sayedihashimi can you re-open?  Like @mredman I found the following problems:

1. When I add the T4MVC nuget package the `.tt` doesn't get copied to the project (I guess this is due to the changes in how NuGet handles content files?  I will open a separate issue on [T4MVC](https://github.com/T4MVC/T4MVC) to track this).
2. When I manually copy the `.tt` file over, right click and choose "Run Custom Tool" VS hangs.

I pushed a sample repro here: https://github.com/kevinkuszyk/aspnet-core-t4mvc

/cc @davidebbo

**Comment by brightertools** [comment]
I have tried running T4 templates to generate multiple files again on a class library package use a new install of windows 10 and VS2015: it still hangs, but the menu items seem to be working (ie running the template  as it hangs when I clicked click one of the)

**Comment by sayedihashimi** [comment]
@kevinkuszyk I have opened a new issue for a more generic implementation for single file generators at https://github.com/aspnet/Tooling/issues/394

>When I add the T4MVC nuget package the .tt doesn't get copied to the project (I guess this is due to the changes in how NuGet handles content files? I will open a separate issue on T4MVC to track this).

This is by design from the runtime. For dnx/dotnet project NuGet packages with content are not supported. To discuss this please open an issue in either http://github.com/aspnet/dnx/ or https://github.com/dotnet/corefx.

>When I manually copy the .tt file over, right click and choose "Run Custom Tool" VS hangs.

Can you open a seperate issue at https://github.com/aspnet/Tooling/issues with specific repro steps?

@brightertools see my comment right above this.

**Comment by giancarloa** [comment]
+1 for T4 support...

yikes... ever since asp.net core hit rc status I began looking at it... almost fainted when i found out about t4... pretty much agree with everything @mcquiggd has said... thanks @mcquiggd for your voice...

**Comment by kevinkuszyk** [comment]
@sayedihashimi thanks.  New issues open as requested.
